File: Part1._Enumeration_problem/Lesson3._Backtracking_algorithm/Arrangement.pas
program Arrangement;
const
    InputFile = 'ARRANGES.INP';
    OutputFile = 'ARRANGES.OUT';
    max = 20;
var
    x: array[1..max] of Integer;
    c: array[1..max] of Boolean;
    n, k: Integer;
    f: Text;

procedure PrintResult; {Thủ tục in cấu hình tìm được}
var
    i: Integer;
begin
    for i := 1 to k do Write(f, x[i], ' ');
    WriteLn(f);
end;

procedure Try(i: Integer); {Thử các cách chọn x[i]}
var
    j: Integer;
begin
    for j := 1 to n do
    if c[j] then {Chỉ xét những giá trị j còn tự do}
        begin
            x[i] := j;
            if i = k then PrintResult {Nếu đã chọn được đến xk thì chỉ việc in kết quả}
            else
                begin
                    c[j] := False; {Đánh dấu: j đã bị chọn}
                    Try(i + 1); {Thủ tục này chỉ xét những giá trị còn tự do gán cho x[i+1]}
                    c[j] := True; {Bỏ đánh dấu: j lại là tự do, bởi sắp tới sẽ thử một cách chọn khác của x[i]}
                end;
        end;
end;

begin
    Assign(f, InputFile); Reset(f);
    ReadLn(f, n, k);
    Assign(f, OutputFile); Rewrite(f);
    FillChar(c, SizeOf(c), True); {Tất cả các số đều chưa bị chọn}
    Try(1); {Thử các cách chọn giá trị của x[1]}
    Close(f);
end.

File: Part1._Enumeration_problem/Lesson3._Backtracking_algorithm/Arrangement.c
#include <stdio.h>
#include <stdbool.h>

int n = 0, k = 0;
int x[21];
bool c[21];
FILE *fi = NULL, *fo = NULL;

void printResult() // Xuất cấu hình hiện tại
{
    for (int i = 1; i <= k; i++)
    {
        fprintf(fo, "%d ", x[i]);
    }
    fprintf(fo, "\n");
}

void Try(int i) // Thử các cách chọn cho x[i]
{
    for (int j = 1; j <= n; j++)
    {
        if (c[j]) // Chỉ xét những giá trị j còn tự do
        {
            x[i] = j;
            if (i == k) // Nếu chọn đến được x[k] thì chỉ cần in kết quả
            {
                printResult();
            }
            else
            {
                c[j] = false; // Đánh dấu j đã được chọn
                Try(i + 1);   // Xét những giá trị j còn tự do để gán cho x[i+1]
                c[j] = true;  // Bỏ đánh dấu: j lại là tự do, bởi sắp tới sẽ thử một cách chọn khác của x[i]
            }
        }
    }
}

int main()
{
    fi = fopen("ARRANGES.INP", "r");
    if (fi != NULL)
    {
        fscanf(fi, "%d %d", &n, &k);
        fclose(fi);
        if (k > 0 && k <= n && n <= 20)
        {
            fo = fopen("ARRANGES.OUT", "w");
            for (int i = 1; i <= n; i++)
            {
                c[i] = true; // Đánh dấu tất cả các số đều chưa được chọn
            }
            Try(1); // Thử các cách chọn cho x[1]
            fclose(fo);
        }
    }
    return 0;
}